src/rgw/rgw_crypt.cc

```diff
@@ -1397,20 +1461,20 @@ int rgw_s3_prepare_decrypt(struct req_state* s,
     /* try to retrieve actual key */
     std::string key_id = get_str_attribute(attrs, RGW_ATTR_CRYPT_KEYID);
     std::string actual_key;
-    res = reconstitute_actual_key_from_sse_s3(s->cct, attrs, actual_key);
+    res = reconstitute_actual_key_from_sse_s3(s, s->cct, attrs, actual_key);
     if (res != 0) {
-      ldpp_dout(s, 10) << "ERROR: failed to retrieve actual key  " << dendl;
-      s->err.message = "Failed to retrieve the actual key " ;
+      ldpp_dout(s, 10) << "ERROR: failed to retrieve actual key" << dendl;
+      s->err.message = "Failed to retrieve the actual key";
       return res;
     }
     if (actual_key.size() != AES_256_KEYSIZE) {
       ldpp_dout(s, 0) << "ERROR: key obtained " <<
-          " is not 256 bit size" << dendl;
-      s->err.message = "SSE-S3  provided an invalid key for the given keyid.";
+          "is not 256 bit size" << dendl;
+      s->err.message = "SSE-S3 provided an invalid key for the given keyid.";
       return -ERR_INVALID_ACCESS_KEY;
     }
 
-    auto aes = std::unique_ptr<AES_256_CBC>(new AES_256_CBC(s->cct));
+    auto aes = std::unique_ptr<AES_256_CBC>(new AES_256_CBC(s, s->cct));
```

Comment:
this is in the op-path, right?  is this something that needs heap allocation?

Reply:
the same idiom is all over rgw_crypt.  if it's wrong it should be addressed all at once, probably separately from this PR.  Looks like there are some datatype visibility issues that would have to be resolved to combine storage allocations like you suggest.